src/Middleware/RequestDecompression/src/RequestDecompressionMiddleware.cs

```diff
@@ -0,0 +1,63 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.AspNetCore.Http;
+
+namespace Microsoft.AspNetCore.RequestDecompression;
+
+/// <summary>
+/// Enables HTTP request decompression.
+/// </summary>
+public class RequestDecompressionMiddleware
+{
+    private readonly RequestDelegate _next;
+    private readonly IRequestDecompressionProvider _provider;
+
+    /// <summary>
+    /// Initialize the request decompression middleware.
+    /// </summary>
+    /// <param name="next">The delegate representing the remaining middleware in the request pipeline.</param>
+    /// <param name="provider">The <see cref="IRequestDecompressionProvider"/>.</param>
+    public RequestDecompressionMiddleware(
+        RequestDelegate next,
+        IRequestDecompressionProvider provider)
+    {
+        if (next == null)
+        {
+            throw new ArgumentNullException(nameof(next));
+        }
+
+        if (provider == null)
+        {
+            throw new ArgumentNullException(nameof(provider));
+        }
+
+        _next = next;
+        _provider = provider;
+    }
+
+    /// <summary>
+    /// Invoke the middleware.
+    /// </summary>
+    /// <param name="context">The <see cref="HttpContext"/>.</param>
+    /// <returns>A task that represents the execution of this middleware.</returns>
+    public Task Invoke(HttpContext context)
+    {
+        if (_provider.ShouldDecompressRequest(context)
+            && _provider.IsContentEncodingSupported(context))
+        {
+            return InvokeCore(context);
+        }
+
+        return _next(context);
+    }
+
+    private async Task InvokeCore(HttpContext context)
+    {
+        var decompressionBody = new RequestDecompressionBody(context, _provider);
+        context.Request.Body = decompressionBody;
+
+        await _next(context);
```

Comment:
Does `IHttpRequestFeature` need to be replaced?

https://github.com/dotnet/aspnetcore/blob/ac27ce1905e239b504b7ffb170c2f5726c3b282f/src/Http/Http.Features/src/IHttpRequestFeature.cs#L88

Reply:
Why replace the feature? Updating the body (and reverting it afterwards) should be sufficient.